### PR TITLE
directory support for wiki

### PR DESCRIPTION
**Pull request by lucastucious**
forked at PR  #7390  

This PR
* add support for directories inside wiki. 
* Create/update/delete pages works well. If an old page is updated, the file will be moved to subdirectory.
* allows to read files with nearly all letters, as it is possible to push all via wiki gitrepo
            files can exists in old and new structure.
                files are selected by link, so it is possible to access and view both
                writing to one of them will delete both, and create new one in directory structure
                delete will delete the selected file
        support images relative to directory path
            fix preview on preview rendering
        fix 500 error if wiki contains pages with invalid uri encoding (e.g. filename: badescaping%%.md)
    ui fixes
        add button 'Abort' to wiki edit and new page

### CODE REVIEW

File: models/wiki.go [hunk 1]
```diff
@@ -39,17 +50,53 @@ func WikiNameToFilename(name string) string {
...
 }
 
+// WikiNameToPathFilename converts a wiki name to its corresponding filename, keep directory paths.
+func WikiNameToPathFilename(name string) string {
+	var restore = [1][2]string{
+		{`(\.\.\/)`, ""}, // remove path up
+	}
+	for _, kv := range restore {
+		loopRe := regexp.MustCompile(kv[0])
+		name = loopRe.ReplaceAllString(name, kv[1])
```
Comment (guillep2k) [inline models/wiki.go:60]: See my other comments above.

File: models/wiki.go [hunk 2]
```diff
@@ -39,17 +50,53 @@ func WikiNameToFilename(name string) string {
...
+	}
+	name = strings.Trim(name, "\n\r\t ./") // trim whitespace and / .
+	return name + ".md"
+}
+
+// FilenameToPathFilename converts a wiki filename to filename with filepath.
+func FilenameToPathFilename(name string) string {
```
Comment (guillep2k) [inline models/wiki.go:67]:
```
import "net/url"
```
And you have `net.PathEscape()`, `net.PathUnescape()`.

File: models/wiki.go [hunk 3]
```diff
@@ -136,6 +198,9 @@ func (repo *Repository) updateWikiPage(doer *User, oldWikiName, newWikiName, con
 	}
 
 	gitRepo, err := git.OpenRepository(basePath)
+
+	fmt.Println(reflect.TypeOf(gitRepo))
```
Comment (guillep2k) [inline models/wiki.go:202]: This should not be here

**Review by guillep2k — COMMENTED** [review]
I've left some comments about the code (only the first couple of files).

The problem I see here (if I understand your intention correctly) is that existing wikis might become invalid after this code is implemented. And a migration process will be extremely difficult, since wikis contain links/references inside the pages and they must match the names of the documents (and page history will become useless).

Perhaps this could be implemented as optional for new repos, though, as those will contain no history.

**Comment by Cherrg** [comment]
> The problem I see here (if I understand your intention correctly) is that existing wikis might become invalid after this code is implemented. And a migration process will be extremely difficult, since wikis contain links/references inside the pages and they must match the names of the documents (and page history will become useless).

Yeah my intention was to support both, old links and new ones (if i remember correctly) -> so old links and paths should still be working. (I hope i don't missed too much cases)
Previously there is a bug with relative paths. May some one with an existing wiki could test this PR.

I will fix my own branch in the next few days so it will compile again. (if @lucasdlk isn't faster)

**Pull request closed without merging**